src/main/java/com/amazon/opendistroforelasticsearch/sql/rewriter/matchtoterm/TermFieldRewriter.java

```diff
@@ -210,19 +214,53 @@ public boolean isValidIdentifierForTerm(SQLIdentifierExpr expr) {
             );
     }
 
-    public boolean hasIdenticalMappings(TermFieldScope scope, Map<String, String> indexToType) {
+    public void checkMappingCompatibility(TermFieldScope scope, Map<String, String> indexToType) {
         if (scope.getMapper().isEmpty()) {
             throw new VerificationException("Unknown index " + indexToType.keySet());
         }
+        // Collect all FieldMappings into hash set and ignore index/type names. set.size > 1 means FieldMappings NOT unique,
+        // needs further compatibility check
+        Set<FieldMappings> set = curScope().getMapper().allMappings().stream().
+                                 flatMap(typeMappings -> typeMappings.allMappings().stream()).
+                                 collect(Collectors.toSet());
 
-        if (isMappingOfAllIndicesDifferent()) {
-            return false;
-        }
+        FieldMappings fieldMappings = null;
+
+        if (set.size() > 1) {
+            Map<String, Map<String, Object>> map = new LinkedHashMap<>();
+
+            for (FieldMappings f : set) {
+                Map<String, Map<String, Object>> m = f.data();
+
+                for (Map.Entry<String, Map<String, Object>> e : m.entrySet()) {
+                    String fieldName = e.getKey();
+                    Map<String, Object> fieldMappingValue = e.getValue();
+
+                    if (!map.containsKey(fieldName)) {
+                        map.put(fieldName, fieldMappingValue);
+                    } else {
+                        // check if types are same
+                        if (!fieldMappingValue.equals(map.get(fieldName))) {
+                           // TODO: Merge mappings if they are compatible, like text and text/keyword to text/keyword.
```

Comment:
I think merge should take the intersection, not the union. Because otherwise one can use properties of a mapping for an index that does not actually have it. Also, please open an issue to track it, I don't think we have a clearly defined process for resolving TODOs.

Reply:
The todo here is only dealing with one case I met while testing, which is the text(or other type) and text(or other type).keyword. It doesn't affect solving this issue. Sure, will create another issue to track it.